lib/Dialect/Arc/Transforms/ArcCanonicalizer.cpp

```diff
@@ -577,6 +583,63 @@ CompRegCanonicalizer::matchAndRewrite(seq::CompRegOp op,
   return success();
 }
 
+LogicalResult
+MergeVectorizeOps::matchAndRewrite(VectorizeOp op,
+                                   PatternRewriter &rewriter) const {
+  auto operandsVector = op.getOperands();
+  auto groupSize = op.getResults().size();
+
+  SmallVector<SmallVector<Value>> operandsGroups;
+  unsigned counter = 0;
+  SmallVector<Value> temp;
+  for (auto operand : operandsVector) {
+    if (counter == groupSize){
+      operandsGroups.push_back(temp);
+      counter = 0;
+      temp.clear();
+    }
+
+    temp.push_back(operand);
+    counter++;
+  }
+
+  for (auto group : operandsGroups) {
+    auto defOp = group[0].getDefiningOp();
+    bool allFromTheSameGroup = true;
+    SmallPtrSet<Value, 32> uniqueOperands;
+    for (auto operand : group) {
+      if (operand.getDefiningOp() != defOp) {
+        allFromTheSameGroup = false;
+        break;
+      }
+
+      if (!operand.hasOneUse())
+        return failure();
+
+      uniqueOperands.insert(operand);
+    }
+
+    if (allFromTheSameGroup) {
+      if (uniqueOperands.size() != groupSize)
+        return failure();
+
+      // Here, we are sure that uniqueOperands' values are from the same
+      // VectorizeOp and they are forming a full vector parameter.
+      
+      // Now we need to pull the `defOp`'s body into `op`'s body
+      // approach:
+      // get the operands of the `defOp` and devide them by two
+      // now we have LHS and RHS, pass the LHS and RHS as a parameter
+      // to the newOp (op) do the calculation (is there a way to know it?)
+      // after this do the calculation of the current op (the op we pulled
+      // the first op into) and boooom! we did it!!
+      // will this work??? am I on the right path so far???
```

Comment:
@TaoBi22 @fabianschuiki @maerhart Can `RewriterBase::inlineRegionBefore` help with this?
or `RewriterBase::inlineBlockBefore` is better?

Reply:
Yeah the inlining should be able to do this I believe. You might have to set up an `IRMapping` with mappings from the other vector's block args to the block args in the vector you're merging into.